### DISCUSSION TITLE
Is it possible to use an existing full Ubuntu OS as a bootstrap agent?

### DISCUSSION THREAD

**Discussion by Heng-Zhou** [sylabs/singularity#2559] [Q&A] [answered]
Suppose we have a machine on which a fresh full Ubuntu 22.04 server or desktop is installed from ISO image file download from Ubuntu official website. Is there any chance we can use the existing whole full Ubuntu installation as a bootstrap agent, or can we make a Singularity container by duplicating the existing whole full Ubuntu installation and then use it as a bootstrap agent? Thank you.

**Comment by dtrudg** [comment, marked as answer]
This is technically possible (creating a container filesystem image of a full system), but the result is not likely not be practical. You could make an image of your entire host system using e.g. `mksquashfs` and then build `Bootstrap: localimage` from that image, but the resulting container is very unlikely to work as you expect.

Because containers do not boot a kernel, and cannot start an init system (systemd) without workarounds, a large amount of configuration and services that are applied / run on the host system will not be applied / run in a container. Software installed in the host system is likely to depend on configuration / services that are then not correct or running in the container environment. Singularity starts containers, by default, in a way where environment variables and mounts are propogated into a container. This is also likely to conflict with the expectations of a 'full system' container.

If you want a clone of a host system, it would be better to use a Virtual Machine approach instead of containers.

As is discussed in the documentation, Singularity intends to allow packaging of applications / pieces of software into a container, not entire systems. We don't intend to document, or introduce any features related to containerizing a full system.

**Comment by Heng-Zhou** [comment]
I would like to confirm, if we use
```
Bootstrap: library
From: ubuntu
```
as header, is the base operating system of guest Ubuntu the _Minimal Ubuntu_ described [here](https://wiki.ubuntu.com/Minimal)?

**Reply by dtrudg** [reply]
The Ubuntu images in the SCS library are SIF conversions of the public Docker images.

If you want to bootstrap Ubuntu from scratch, I suggest using the `debootstrap` bootstrap, as in the examples:

https://github.com/sylabs/singularity/blob/main/examples/ubuntu/Singularity

https://docs.sylabs.io/guides/4.1/user-guide/appendix.html#debootstrap-build-agent

**Comment by Heng-Zhou** [comment]
Thank you for the clarification.
